File: backend/src/core/rashi.py
"""
Rashi (Zodiac Sign) calculations for Vedic Astrology.
Each rashi spans exactly 30 degrees of the sidereal zodiac.
"""

# Rashis in Sanskrit with English equivalents
RASHIS = [
    'Mesha',      # Aries
    'Vrishabha',  # Taurus
    'Mithuna',    # Gemini
    'Karka',      # Cancer
    'Simha',      # Leo
    'Kanya',      # Virgo
    'Tula',       # Libra
    'Vrischika', # Scorpio
    'Dhanu',      # Sagittarius
    'Makara',     # Capricorn
    'Kumbha',     # Aquarius
    'Meena'       # Pisces
]

# English names for reference
RASHI_ENGLISH = [
    'Aries', 'Taurus', 'Gemini', 'Cancer',
    'Leo', 'Virgo', 'Libra', 'Scorpio',
    'Sagittarius', 'Capricorn', 'Aquarius', 'Pisces'
]

# Rashi lords
RASHI_LORDS = {
    'Mesha': 'Mars',
    'Vrishabha': 'Venus',
    'Mithuna': 'Mercury',
    'Karka': 'Moon',
    'Simha': 'Sun',
    'Kanya': 'Mercury',
    'Tula': 'Venus',
    'Vrischika': 'Mars',
    'Dhanu': 'Jupiter',
    'Makara': 'Saturn',
    'Kumbha': 'Saturn',
    'Meena': 'Jupiter'
}

# Rashi elements
RASHI_ELEMENTS = {
    'Mesha': 'Fire',
    'Vrishabha': 'Earth',
    'Mithuna': 'Air',
    'Karka': 'Water',
    'Simha': 'Fire',
    'Kanya': 'Earth',
    'Tula': 'Air',
    'Vrischika': 'Water',
    'Dhanu': 'Fire',
    'Makara': 'Earth',
    'Kumbha': 'Air',
    'Meena': 'Water'
}

# Rashi modalities
RASHI_MODALITIES = {
    'Mesha': 'Movable',      # Chara
    'Vrishabha': 'Fixed',    # Sthira
    'Mithuna': 'Dual',       # Dwiswabhava
    'Karka': 'Movable',
    'Simha': 'Fixed',
    'Kanya': 'Dual',
    'Tula': 'Movable',
    'Vrischika': 'Fixed',
    'Dhanu': 'Dual',
    'Makara': 'Movable',
    'Kumbha': 'Fixed',
    'Meena': 'Dual'
}


def get_rashi(sidereal_longitude: float) -> tuple[int, str, float]:
    """
    Get rashi (zodiac sign) from sidereal longitude.
    
    Args:
        sidereal_longitude: Longitude in sidereal zodiac (0-360)
        
    Returns:
        tuple: (rashi_index, rashi_name, degree_in_rashi)
        - rashi_index: 0-11 (Mesha to Meena)
        - rashi_name: Sanskrit name of the rashi
        - degree_in_rashi: Degree within the rashi (0-30)
    """
    # Normalize longitude to 0-360 range
    longitude = sidereal_longitude % 360
    
    # Each rashi spans exactly 30 degrees
    rashi_index = int(longitude / 30)
    degree_in_rashi = longitude % 30
    
    return rashi_index, RASHIS[rashi_index], degree_in_rashi


def get_rashi_info(rashi_index: int) -> dict:
    """
    Get detailed information about a rashi.
    
    Args:
        rashi_index: Index of the rashi (0-11)
        
    Returns:
        Dictionary with rashi details
    """
    name = RASHIS[rashi_index]
    return {
        'index': rashi_index,
        'name': name,
        'english': RASHI_ENGLISH[rashi_index],
        'lord': RASHI_LORDS[name],
        'element': RASHI_ELEMENTS[name],
        'modality': RASHI_MODALITIES[name]
    }


def get_opposite_rashi(rashi_index: int) -> int:
    """Get the opposite rashi (180° away)."""
    return (rashi_index + 6) % 12


def get_trine_rashis(rashi_index: int) -> list[int]:
    """Get trine rashis (120° apart) - same element."""
    return [
        rashi_index,
        (rashi_index + 4) % 12,
        (rashi_index + 8) % 12
    ]


def get_square_rashis(rashi_index: int) -> list[int]:
    """Get square rashis (90° apart)."""
    return [
        (rashi_index + 3) % 12,
        (rashi_index + 9) % 12
    ]

File: frontend/src/lib/core/rashi.ts
/** Rashi (Zodiac Sign) calculations — direct port of backend/src/core/rashi.py */

export const RASHIS = [
  'Mesha', 'Vrishabha', 'Mithuna', 'Kataka',
  'Simha', 'Kanya', 'Tula', 'Vrischika',
  'Dhanu', 'Makara', 'Kumbha', 'Meena',
] as const;

export const RASHI_ENGLISH = [
  'Aries','Taurus','Gemini','Cancer',
  'Leo','Virgo','Libra','Scorpio',
  'Sagittarius','Capricorn','Aquarius','Pisces',
] as const;

/** Sinhala (Sanskrit-derived) rashi names, index-aligned with RASHIS. */
export const RASHI_SI = [
  'මේෂ', 'වෘෂභ', 'මිථුන', 'කටක',
  'සිංහ', 'කන්‍යා', 'තුලා', 'වෘශ්චික',
  'ධනු', 'මකර', 'කුම්භ', 'මීන',
] as const;

/** Tamil (Sanskrit-derived) rashi names, index-aligned with RASHIS. */
export const RASHIS_TA = [
  'மேஷம்', 'ரிஷபம்', 'மிதுனம்', 'கடகம்',
  'சிம்மம்', 'கன்னி', 'துலாம்', 'விருச்சிகம்',
  'தனுசு', 'மகரம்', 'கும்பம்', 'மீனம்',
] as const;

/** Chinese zodiac-sign names, the standard terms used regardless of sidereal/tropical framing. */
export const RASHIS_ZH = [
  '白羊座', '金牛座', '双子座', '巨蟹座',
  '狮子座', '处女座', '天秤座', '天蝎座',
  '射手座', '摩羯座', '水瓶座', '双鱼座',
] as const;

/** Hindi (Sanskrit-derived) rashi names, index-aligned with RASHIS. */
export const RASHIS_HI = [
  'मेष', 'वृषभ', 'मिथुन', 'कर्क',
  'सिंह', 'कन्या', 'तुला', 'वृश्चिक',
  'धनु', 'मकर', 'कुंभ', 'मीन',
] as const;

/** Japanese zodiac-sign names — no native Vedic tradition, so the standard Western-derived terms. */
export const RASHIS_JA = [
  '牡羊座', '牡牛座', '双子座', '蟹座',
  '獅子座', '乙女座', '天秤座', '蠍座',
  '射手座', '山羊座', '水瓶座', '魚座',
] as const;

/** Korean zodiac-sign names — no native Vedic tradition, so the standard Western-derived terms. */
export const RASHIS_KO = [
  '양자리', '황소자리', '쌍둥이자리', '게자리',
  '사자자리', '처녀자리', '천칭자리', '전갈자리',
  '사수자리', '염소자리', '물병자리', '물고기자리',
] as const;

/** Arabic zodiac-sign names. */
export const RASHIS_AR = [
  'الحمل', 'الثور', 'الجوزاء', 'السرطان',
  'الأسد', 'العذراء', 'الميزان', 'العقرب',
  'القوس', 'الجدي', 'الدلو', 'الحوت',
] as const;

/** Malayalam rashi names — the same Sanskrit-derived terms Kerala's own calendar months use. */
export const RASHIS_ML = [
  'മേടം', 'ഇടവം', 'മിഥുനം', 'കർക്കടകം',
  'ചിങ്ങം', 'കന്നി', 'തുലാം', 'വൃശ്ചികം',
  'ധനു', 'മകരം', 'കുംഭം', 'മീനം',
] as const;

export const RASHI_LORDS: Record<string, string> = {
  Mesha:'Mars', Vrishabha:'Venus', Mithuna:'Mercury', Kataka:'Moon',
  Simha:'Sun', Kanya:'Mercury', Tula:'Venus', Vrischika:'Mars',
  Dhanu:'Jupiter', Makara:'Saturn', Kumbha:'Saturn', Meena:'Jupiter',
};

export const RASHI_ELEMENTS: Record<string, string> = {
  Mesha:'Fire', Vrishabha:'Earth', Mithuna:'Air', Kataka:'Water',
  Simha:'Fire', Kanya:'Earth', Tula:'Air', Vrischika:'Water',
  Dhanu:'Fire', Makara:'Earth', Kumbha:'Air', Meena:'Water',
};

export const RASHI_MODALITIES: Record<string, string> = {
  Mesha:'Movable', Vrishabha:'Fixed', Mithuna:'Dual', Kataka:'Movable',
  Simha:'Fixed', Kanya:'Dual', Tula:'Movable', Vrischika:'Fixed',
  Dhanu:'Dual', Makara:'Movable', Kumbha:'Fixed', Meena:'Dual',
};

export function getRashi(siderealLon: number): [number, string, number] {
  const lon = ((siderealLon % 360) + 360) % 360;
  const idx = Math.floor(lon / 30);
  return [idx, RASHIS[idx], lon % 30];
}

export function getRashiInfo(rashiIndex: number) {
  const name = RASHIS[rashiIndex];
  return {
    index: rashiIndex,
    name,
    english: RASHI_ENGLISH[rashiIndex],
    lord: RASHI_LORDS[name],
    element: RASHI_ELEMENTS[name],
    modality: RASHI_MODALITIES[name],
  };
}

export function getOppositRashi(idx: number): number { return (idx + 6) % 12; }
export function getTrineRashis(idx: number): number[] { return [idx, (idx+4)%12, (idx+8)%12]; }
export function getSquareRashis(idx: number): number[] { return [(idx+3)%12, (idx+9)%12]; }
